### PR TITLE
SI-6036 yet again makes sense of magic symbols

### PR DESCRIPTION
**Pull request by xeno-by**
review by @odersky

### CODE REVIEW

**Comment by cvogt** [comment]
@odersky this is a blocker for the new resolveOverloaded implementation. Our GSOC student Chris Hodapp is working on resolveOverloaded intensively this weekend, but leaves Monday for vacation during the scheduled GSOC student week off. If we can accept this request soon, he may finish and the new resolveOverloaded gets into M5.

**Comment by gkossakowski** [comment]
Am I the only one that has no understanding of the problem and solution to it introduced by this commit?

**Comment by xeno-by** [comment]
In the compiler there are some magic symbols (outlined in this commit) that don't correspond to any code. They are synthesized on the fly when a symbol table is initialized.

Since the reflection refactoring we can have a hierarchy of symbol tables, however, some parts of the compiler still thought that there's only one. Hence synthetic symbols were entered only into the root symbol table. This commit is the final take on this problem that is supposed to fix it for good.

A brief recap on the employed solution. Symbol synthesis is done in Definitions, which is bound to the root symbol table. Hence it's tough to invoke this logic for other symbol tables (because the root mirror doesn't know when a child symbol table is created). Instead we synthesize magic symbols in the missingHook callback.

**Comment by gkossakowski** [comment]
@xeno-by: Thanks! That makes a perfect sense now. I'd really appreciate that sort of descriptions going into commit messages. I think that's the only way others can eventually help with maintaining the code you are writing.

**Comment by xeno-by** [comment]
Right. I usually do that, but now it was my oversight.

On 9 July 2012 12:08, Grzegorz Kossakowski <
reply@reply.github.com

File: src/reflect/scala/reflect/internal/Definitions.scala [hunk 1]
```diff
@@ -1125,6 +1125,39 @@ trait Definitions extends api.StandardDefinitions {
...
+      Object_ne,
+      Object_==,
+      Object_!=,
+      Object_##,
+      Object_synchronized,
+      Object_isInstanceOf,
+      Object_asInstanceOf,
+      String_+,
+      ComparableClass,
+      JavaSerializableClass
```
Comment (paulp) [inline src/reflect/scala/reflect/internal/Definitions.scala:1159]: Its hard to imagine that most of these symbols need to be on this list.  I know it was taken wholesale from the list of symbols forced at the end of definitions, but that was method-private.  Suddenly it's the canonical list of "magic" symbols.

**Comment by odersky** [comment]
Yes, I agree. Conflates two issues: What needs to be initialized during definitions, and what is a phantom class. I vote for going back and isolating phantom classes only.

**Pull request merged**